swift/aperture/main.swift

```diff
@@ -31,21 +19,35 @@ func record() {
   var coordinates = [String]()
   if cropArea != "none" {
     coordinates = CommandLine.arguments[3].components(separatedBy: ":")
-    if coordinates.count - 1 != 3 { // Number of ':' in the string
-      print("The coordinates for the crop rect must be in the format 'originX:originY:width:height'")
-      exit(2)
-    }
   }
 
-  recorder = Recorder(
-    destinationPath: destinationPath,
-    fps: fps,
-    coordinates: coordinates,
-    showCursor: showCursor,
-    highlightClicks: highlightClicks,
-    displayId: displayId!,
-    audioDeviceId: audioDeviceId
-  )
+  do {
+    recorder = try Recorder(
+      destinationPath: destinationPath,
+      fps: fps,
+      coordinates: coordinates,
+      showCursor: showCursor,
+      highlightClicks: highlightClicks,
+      displayId: displayId!,
+      audioDeviceId: audioDeviceId
+    )
+  } catch {
+    printErr(error)
+    exit(1)
+  }
+
+  recorder.onStart = {
+    print("R")
+  }
+
+  recorder.onFinish = {
+    exit(0)
+  }
+
+  recorder.onError = {
+    printErr($0)
+    exit(1)
+  }
```

Comment:
I like closures and think they work well here! 👍 

With closures you have to be careful not to introduce retain cycles though, and this can be hard sometimes in swift.
With delegates the api can take responsibility and break the retain cycles, so that the caller doesn't have too..

You don't refrerence self or any variables inside the closures so you don't need to worry about that here..

[Good blog post about reference cycles in closures](https://krakendev.io/blog/weak-and-unowned-references-in-swift)